lib/matplotlib/colors.py

```diff
@@ -2066,3 +2074,83 @@ def from_levels_and_colors(levels, colors, extend='neither'):
 
     norm = BoundaryNorm(levels, ncolors=n_data_colors)
     return cmap, norm
+
+
+def darker(color, factor):
+    """
+    A helper routine to to generate a darker shade of color.
+
+    Parameters
+    ----------
+    color : color or color sequence
+        The colors to convert.
+    factor : float
+        any value above 1 means darkening the color,
+        any value in [0,1] means brightening the color
+
+    Returns
+    -------
+    result : (..., 3) ndarray
+       Colors converted to RGB values in range [0, 1]
+
+    Notes
+    -----
+    The algorithm is inspired by Qt ``QColor.darker()``.
+    """
+    if factor <= 0:
+        raise ValueError("Cannot input value below 0.")
+    elif factor < 1:
+        return lighter(color, 1 / factor)
+    color = to_rgb(color)
+    hsv = rgb_to_hsv(color)
+    hsv[:][2] = hsv[:][2] / factor
+
+    return hsv_to_rgb(hsv)
+
+
+def lighter(color, factor):
+    """
+    A helper routine to to generate a lighter shade of color inspired by PyQt.
+
+    Parameters
+    ----------
+    color : color or color sequence
+        The colors to convert.
+    factor : float
+        any value above 1 means brightening the color,
+        any value in [0,1] means darkening the color
+
+    Returns
+    -------
+    result : (..., 3) ndarray
+       Colors converted to RGB values in range [0, 1]
+    """
+
+    if factor <= 0:
+        raise ValueError("Cannot input value below 0.")
+    elif factor < 1:
+        return darker(color, 1 / factor)
+    color = to_rgb(color)
+    hsv = rgb_to_hsv(color)
+    is_scalar = hsv.ndim <= 1
+    h, s, v = np.atleast_2d(hsv).T
+    s = s * 255
+    v = v * 255
+
+    v = (factor * v)
+
+    for i in range(len(v)):
+        if v[i] > 255:
+            s[i] -= v[i] - 255
+            if s[i] < 0:
```

Comment:
This whole block on v should be vectorized and the redundant conversion from/to 255 removed.